File: sdk-go/account_test.go
package boxkite

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"testing"
)

func TestUsage(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/usage" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"monthly_sandbox_hours_used": 1.5, "monthly_sandbox_hours_limit": 20, "concurrent_sandboxes": 1, "concurrent_sandboxes_limit": 3}`)
	})
	defer closeServer()

	usage, err := client.Usage(context.Background())
	if err != nil {
		t.Fatalf("Usage: %v", err)
	}
	if usage.MonthlySandboxHoursUsed != 1.5 || usage.ConcurrentSandboxesLimit != 3 {
		t.Errorf("unexpected usage: %+v", usage)
	}
}

func TestRequestPasswordReset_PostsEmail(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/auth/password-reset/request" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["email"] != "user@example.com" {
			t.Errorf("unexpected email: %v", body["email"])
		}
		writeJSON(t, w, 200, `{"message": "If an account with that email exists, a password reset link has been sent."}`)
	})
	defer closeServer()

	result, err := client.RequestPasswordReset(context.Background(), "user@example.com")
	if err != nil {
		t.Fatalf("RequestPasswordReset: %v", err)
	}
	if result.Message == "" {
		t.Error("expected a non-empty message")
	}
}

func TestConfirmPasswordReset_RaisesOnInvalidToken(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, 400, `{"error": {"code": "invalid_or_expired_token", "message": "This password reset link is invalid or has expired."}}`)
	})
	defer closeServer()

	_, err := client.ConfirmPasswordReset(context.Background(), "bad-tok", "new-hunter2")
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T", err)
	}
	if apiErr.Code != "invalid_or_expired_token" {
		t.Errorf("unexpected code: %s", apiErr.Code)
	}
}

func TestVerifyEmail_PostsToken(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["token"] != "verify-tok" {
			t.Errorf("unexpected token: %v", body["token"])
		}
		writeJSON(t, w, 200, `{"message": "Email verified."}`)
	})
	defer closeServer()

	result, err := client.VerifyEmail(context.Background(), "verify-tok")
	if err != nil {
		t.Fatalf("VerifyEmail: %v", err)
	}
	if result.Message != "Email verified." {
		t.Errorf("unexpected message: %s", result.Message)
	}
}

func TestResendVerification_OverridesAuthorizationWithAccessToken(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if got := r.Header.Get("Authorization"); got != "Bearer dashboard-jwt-123" {
			t.Errorf("expected the dashboard JWT to replace the client's api_key, got %q", got)
		}
		writeJSON(t, w, 200, `{"message": "Verification email sent."}`)
	})
	defer closeServer()

	result, err := client.ResendVerification(context.Background(), "dashboard-jwt-123")
	if err != nil {
		t.Fatalf("ResendVerification: %v", err)
	}
	if result.Message != "Verification email sent." {
		t.Errorf("unexpected message: %s", result.Message)
	}
}

func TestRefreshToken_PostsRefreshTokenAndReturnsNewPair(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["refresh_token"] != "old-refresh" {
			t.Errorf("unexpected refresh_token: %v", body["refresh_token"])
		}
		writeJSON(t, w, 200, `{
			"access_token": "new-jwt", "token_type": "bearer", "expires_in": 3600, "refresh_token": "new-refresh",
			"account": {"id": "acct-1", "email": "a@example.com", "created_at": "2026-01-01T00:00:00Z"}
		}`)
	})
	defer closeServer()

	pair, err := client.RefreshToken(context.Background(), "old-refresh")
	if err != nil {
		t.Fatalf("RefreshToken: %v", err)
	}
	if pair.AccessToken != "new-jwt" || pair.RefreshToken == nil || *pair.RefreshToken != "new-refresh" {
		t.Errorf("unexpected pair: %+v", pair)
	}
}

func TestRefreshToken_RaisesOnReusedToken(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		writeJSON(t, w, 401, `{"error": {"code": "refresh_token_reused", "message": "This refresh token has already been used."}}`)
	})
	defer closeServer()

	_, err := client.RefreshToken(context.Background(), "already-used")
	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("expected *APIError, got %T", err)
	}
	if apiErr.Code != "refresh_token_reused" || apiErr.StatusCode != 401 {
		t.Errorf("unexpected APIError: %+v", apiErr)
	}
}

func TestLogout_PostsRefreshTokenAndReturnsNoError(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		if body["refresh_token"] != "some-refresh" {
			t.Errorf("unexpected refresh_token: %v", body["refresh_token"])
		}
		w.WriteHeader(204)
	})
	defer closeServer()

	if err := client.Logout(context.Background(), "some-refresh"); err != nil {
		t.Fatalf("Logout: %v", err)
	}
}

func TestGetAllowedCommands(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/account/allowed-commands" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		writeJSON(t, w, 200, `{"rules": ["git", {"command": "python3", "args_allow": ["^-c .*"], "args_deny": []}]}`)
	})
	defer closeServer()

	result, err := client.GetAllowedCommands(context.Background())
	if err != nil {
		t.Fatalf("GetAllowedCommands: %v", err)
	}
	if len(result.Rules) != 2 {
		t.Fatalf("expected 2 rules, got %d", len(result.Rules))
	}
	if result.Rules[0].Command != "git" {
		t.Errorf("unexpected first rule: %+v", result.Rules[0])
	}
	if result.Rules[1].Command != "python3" || len(result.Rules[1].ArgsAllow) != 1 {
		t.Errorf("unexpected second rule: %+v", result.Rules[1])
	}
}

func TestSetAllowedCommands_SendsRulesAndReturnsEcho(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPut || r.URL.Path != "/v1/account/allowed-commands" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		raw, _ := io.ReadAll(r.Body)
		var body map[string]any
		_ = json.Unmarshal(raw, &body)
		rules, _ := body["rules"].([]any)
		if len(rules) != 1 {
			t.Fatalf("expected 1 rule, got %v", body["rules"])
		}
		writeJSON(t, w, 200, `{"rules": [{"command": "git", "args_allow": [], "args_deny": []}]}`)
	})
	defer closeServer()

	result, err := client.SetAllowedCommands(context.Background(), []AllowedCommandRule{{Command: "git"}})
	if err != nil {
		t.Fatalf("SetAllowedCommands: %v", err)
	}
	if len(result.Rules) != 1 || result.Rules[0].Command != "git" {
		t.Errorf("unexpected result: %+v", result)
	}
}

func TestClearAllowedCommands(t *testing.T) {
	client, closeServer := newTestClient(t, func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete || r.URL.Path != "/v1/account/allowed-commands" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(204)
	})
	defer closeServer()

	if err := client.ClearAllowedCommands(context.Background()); err != nil {
		t.Fatalf("ClearAllowedCommands: %v", err)
	}
}

File: sdk-rust/tests/account_test.rs
mod common;

use boxkite_client::AllowedCommandRule;
use serde_json::json;
use wiremock::matchers::{body_json, method, path};
use wiremock::{Mock, ResponseTemplate};

#[tokio::test]
async fn account_returns_identity() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/account"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "id": "acct_1", "email": "dev@example.com", "created_at": "2026-01-01T00:00:00Z"
        })))
        .mount(&server)
        .await;

    let account = client.account().await.expect("account should succeed");
    assert_eq!(account.id, "acct_1");
    assert_eq!(account.email, "dev@example.com");
}

#[tokio::test]
async fn usage_returns_fair_use_limits() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/usage"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "monthly_sandbox_hours_used": 1.5,
            "monthly_sandbox_hours_limit": 100.0,
            "concurrent_sandboxes": 2,
            "concurrent_sandboxes_limit": 10
        })))
        .mount(&server)
        .await;

    let usage = client.usage().await.expect("usage should succeed");
    assert_eq!(usage.concurrent_sandboxes, 2);
    assert_eq!(usage.monthly_sandbox_hours_limit, 100.0);
}

#[tokio::test]
async fn request_password_reset_sends_email() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/auth/password-reset/request"))
        .and(body_json(json!({"email": "dev@example.com"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "message": "If that email is registered, a reset link has been sent."
        })))
        .mount(&server)
        .await;

    let resp = client
        .request_password_reset("dev@example.com")
        .await
        .expect("request should succeed");
    assert!(resp.message.contains("reset link"));
}

#[tokio::test]
async fn refresh_token_returns_new_pair() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/auth/refresh"))
        .and(body_json(json!({"refresh_token": "old_refresh"})))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "access_token": "new_access",
            "token_type": "bearer",
            "expires_in": 3600,
            "refresh_token": "new_refresh",
            "account": {"id": "acct_1", "email": "dev@example.com", "created_at": "2026-01-01T00:00:00Z"}
        })))
        .mount(&server)
        .await;

    let pair = client
        .refresh_token("old_refresh")
        .await
        .expect("refresh should succeed");
    assert_eq!(pair.access_token, "new_access");
    assert_eq!(pair.refresh_token.as_deref(), Some("new_refresh"));
    assert_eq!(pair.account.id, "acct_1");
}

#[tokio::test]
async fn resend_verification_uses_dashboard_jwt_not_api_key() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/auth/resend-verification"))
        .and(wiremock::matchers::header(
            "authorization",
            "Bearer dashboard_jwt_abc",
        ))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({"message": "sent"})))
        .mount(&server)
        .await;

    let resp = client
        .resend_verification("dashboard_jwt_abc")
        .await
        .expect("resend should succeed");
    assert_eq!(resp.message, "sent");
}

#[tokio::test]
async fn logout_sends_refresh_token_and_ignores_empty_body() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("POST"))
        .and(path("/v1/auth/logout"))
        .and(body_json(json!({"refresh_token": "r1"})))
        .respond_with(ResponseTemplate::new(204))
        .mount(&server)
        .await;

    client.logout("r1").await.expect("logout should succeed");
}

#[tokio::test]
async fn get_allowed_commands_decodes_bare_and_object_rules() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("GET"))
        .and(path("/v1/account/allowed-commands"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "rules": [
                "ls",
                {"command": "git", "args_allow": ["^status$"], "args_deny": ["^push$"]}
            ]
        })))
        .mount(&server)
        .await;

    let resp = client
        .get_allowed_commands()
        .await
        .expect("get should succeed");
    assert_eq!(resp.rules.len(), 2);
    assert_eq!(resp.rules[0].command, "ls");
    assert!(resp.rules[0].args_allow.is_empty());
    assert_eq!(resp.rules[1].command, "git");
    assert_eq!(resp.rules[1].args_allow, vec!["^status$".to_string()]);
    assert_eq!(resp.rules[1].args_deny, vec!["^push$".to_string()]);
}

#[tokio::test]
async fn set_allowed_commands_serializes_object_form() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("PUT"))
        .and(path("/v1/account/allowed-commands"))
        .and(body_json(json!({
            "rules": [
                {"command": "ls"},
                {"command": "git", "args_allow": ["^status$"]}
            ]
        })))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({
            "rules": [{"command": "ls"}, {"command": "git", "args_allow": ["^status$"]}]
        })))
        .mount(&server)
        .await;

    let rules = vec![
        AllowedCommandRule::new("ls"),
        AllowedCommandRule::new("git").args_allow(["^status$"]),
    ];
    let resp = client
        .set_allowed_commands(rules)
        .await
        .expect("set should succeed");
    assert_eq!(resp.rules.len(), 2);
}

#[tokio::test]
async fn clear_allowed_commands_sends_delete() {
    let server = common::mock_server().await;
    let client = common::client_for(&server);

    Mock::given(method("DELETE"))
        .and(path("/v1/account/allowed-commands"))
        .respond_with(ResponseTemplate::new(204))
        .mount(&server)
        .await;

    client
        .clear_allowed_commands()
        .await
        .expect("clear should succeed");
}
